core/sched/rt-rm.c

```diff
@@ -0,0 +1,233 @@
+#include <core/sched/scheduler_skeleton.h>
+#include <core/sched/sched-config.h>
+#include <config.h>
+#include <arch/armv7.h>
+#include <stdlib.h>
+#include <debug.h>
+#include <stdbool.h>
+
+#include <lib/list.h>
+#include <core/timer.h>
+
+// #define SCHED_RM_REPORT
+
+struct entry_data_rm {
+    struct sched_entry *e;
+
+    uint32_t period;
+    uint32_t budget;
+
+    uint32_t period_cntdn;
+    uint32_t budget_cntdn;
+
+#ifdef SCHED_RM_REPORT
+    /* for analysis */
+    uint32_t foreground_ticks;
+    uint32_t preempted;
+#endif
+
+    struct list_head head;
+};
+
+struct sched_data_rm {
+    struct scheduler *s;
+
+    uint64_t tick_interval_us;
+    uint64_t tick_count;
+#ifdef SCHED_RM_REPORT
+    uint64_t tick_1kcount;
+#endif
+
+    struct entry_data_rm *current;
+    struct list_head runqueue;
+};
+
+/* Scheduler related data initialization */
+void sched_rm_init(struct scheduler *s)
+{
+    struct sched_data_rm *sd = (struct sched_data_rm *) malloc(sizeof(struct sched_data_rm));
```

Comment:
You added lots of data structures for scheduling, but these are very difficult to read your code, we need to make those things intuitively.